src/rulesets/Base/indexing.jl

```diff
@@ -54,38 +54,77 @@ end
 
 
 #####
-##### getindex
+##### getindex(::AbstractArray)
 #####
 
 function frule((_, ẋ), ::typeof(getindex), x::AbstractArray, inds...)
     return x[inds...], ẋ[inds...]
 end
 
-function rrule(::typeof(getindex), x::Array{<:Number}, inds...)
+function rrule(::typeof(getindex), x::AbstractArray, inds...)
     # removes any logical indexing, CartesianIndex etc
     # leaving us just with a tuple of Int, Arrays of Int and Ranges of Int
     plain_inds = Base.to_indices(x, inds)
     y = getindex(x, plain_inds...)
     function getindex_pullback(ȳ)
-        function getindex_add!(Δ)
-            # this a optimizes away for simple cases
-            for (ȳ_ii, ii) in zip(ȳ, Iterators.product(plain_inds...))
-                Δ[ii...] += ȳ_ii
-            end
-            return Δ
-        end
-
-        x̄ = InplaceableThunk(
-            getindex_add!,
-            @thunk(getindex_add!(zero(x))),
+        xthunk = InplaceableThunk(
+            x̄ -> ∇getindex!(x̄, x, unthunk(ȳ), plain_inds...),
+            @thunk(∇getindex(x, unthunk(ȳ), plain_inds...)),
         )
-        īnds = broadcast(Returns(NoTangent()), inds)
-        return (NoTangent(), x̄, īnds...)
+        nots = map(Returns(NoTangent()), inds)
+        return (NoTangent(), xthunk, nots...)
     end
-
     return y, getindex_pullback
 end
 
+"""
+    ∇getindex(x, dy, inds...)
+
+For the `rrule` of `y = x[inds...]`, this function is roughly 
+`setindex(zero(x), dy, inds...)`, returning the array `dx`.
+Differentiable. Includes `ProjectTo(x)(dx)`.
+"""
+function ∇getindex(x::AbstractArray{<:Number}, dy, inds...)
+    # It's unfortunate to close over `x`, but `similar(typeof(x), axes(x))` doesn't 
+    # allow `eltype(dy)`, nor does it work for many structured matrices.
+    dx = fill!(similar(x, eltype(dy), axes(x)), 0)
+    ∇getindex!(dx, x, dy, inds...)
+    return ProjectTo(x)(dx)  # since we have x, may as well do this inside, not in rules
+end
+function ∇getindex(x::AbstractArray, dy, inds...)
+    # Since we have `x`, we can also handle arrays of arrays.
+    dx = map(zero, x)  # this ignores type of dy, TODO?
+    ∇getindex!(dx, x, dy, inds...)
+    return ProjectTo(x)(dx)
+end
+
+function ∇getindex!(dx::AbstractArray, x::AbstractArray, dy, inds::Integer...)
+    view(dx, inds...) .+= Ref(dy)
+    return dx
+end
+function ∇getindex!(dx::AbstractArray, x::AbstractArray, dy, inds...)
+    view(dx, inds...) .+= dy
+    # For GPU arrays, `inds::Union{Integer, Base.Slice}...` is fine, but any other AbstractArray risks overwriting.
+    # Those should call `NNlib.scatter!`, alla https://github.com/FluxML/Zygote.jl/pull/1131
+    return dx
+end
```

Comment:
I think this function could dispatch to call `scatter!` for sufficiently complicated indexing, when acting on AbstractGPUArray. Without this, the gradient of things like `A[[1,2,2,2]]` will lead to race conditions & wrong answers. But where such an overload should live seems unclear, maybe NNlibCUDA needs to load ChainRules? Or the stub of `∇getindex!` could live in ChainRulesCore?

Reply:
Possibly `scatter!` could live in some 3rd library that defines it for both normal array and GPU arrays and then ChainRules.jl could use that?

Reply:
Right now that is NNlib.jl. Not sure how ChainRules feels about depending on that. GPU version is in NNlibCUDA.jl, which isn't loaded automatically.

Reply:
Reading https://github.com/FluxML/Zygote.jl/pull/1131 again, the winner was in fact copying back to the CPU. I think that could be done with only GPUArraysCore. But for now, this is no worse than Zygote, so perhaps we can postpone fixing that.